computeStudentLoan.ts

```diff
@@ -0,0 +1,141 @@
+import moment from 'moment';
+/**
+* A student loan that accrues interest over time.
+* pr
+*/
+export interface Loan {
+    /** The current decimal principal of the balance in USD. */
+    principal: string;
+    /**
+    * APR interest rate for loan.
+    *
+    * Expressed per-unit (e.g: 6.3% is 0.063).
+    */
+    interestRate: string;
+    /**
+    * Minimum payment required monthly for the loan.
+    *
+    * This establishes the term left in the loan.
+    */
+    minimumPayment: string;
+}
+/**
+* A principal at a given point in time.
+*/
+export interface DateAndPrincipal {
+    /**
+    * Unix epoch milliseconds for when a transaction is effectively
+    * registered at.
+    *
+    * Might be binned to date in some cases.
+    */
+    date: number;
+    /**
+    * Amount transacted at dateCreated.
+    *
+    * Might be a summation in binning cases.
+    *
+    * Represented as a string to have consistent currency representation.
+    */
+    principal: number;
+}
+/**
+* A pair of a loan and the principal values calculated over time.
+*/
+export interface LoanAndPrincipal {
+    /**
+    * Loan these principals are valid for.
+    */
+    loan: Loan;
+    /**
+    * Series of principal values over time.
+    */
+    principal: DateAndPrincipal[];
+}
+/**
+* Given a series of loans, computes the principal over time.
+*
+* Optimizes payments prioritizing any excess payments to the highest
+* interest rate loan.
+*
+* @param loans The loans to pay down.
+* @param monthlyPayment The total payment available to distribute
+* across loans.
+*/
+export function loanPrincipalsOverTime(
+    loans: Loan[], monthlyPayment: number): LoanAndPrincipal[] {
+    if (loans.length === 0) {
+        return [];
+    }
+    const loanAndPrincipals = loans.map((loan) => ({
+        loan: loan,
+        principal: [{
+            date: currentDate.valueOf(),
+            principal: parseFloat(loan.principal),
+        }],
+    }));
+    // Prioritize payments by interest rate.
+    loanAndPrincipals.sort((left, right) => {
+        return (latestPrincipal(left).principal
+            - latestPrincipal(right).principal);
+    });
+    const currentDate = moment().add(1, 'month');
+    // Iterate while loans have balance.
+    while (hasBalance(loanAndPrincipals)) {
```

Comment:
this `while` loop iterates over the same array many times
the complexity can be represented as O(log2(n)^2) * 4 * months-projected
4 internal loops multiplies the complexity by 4, and each of those is squared because the `while` loop has a condition that is itself a loop. It is again multiplied by the number of months projected. This can be greatly simplified. consider separating into a recursive function such as:
```ts
    const iteratWithBalance = (loanAndPrincipal: LoanAndPrincipal, residualPayment: number): LoanAndPrincipal => {
        // Iterate while loans have balance.
        const latest = latestPrincipal(loanAndPrincipal).principal;

        if (latest > 0) {
            const minPayment = Math.min(residualPayment, parseFloat(loanAndPrincipal.loan.minimumPayment));
            residualPayment -= minPayment;
            const extraPayment = Math.min(residualPayment, latest)
            residualPayment -= extraPayment;
            const principalWithInterest = latest * (1.0 + parseFloat(loanAndPrincipal.loan.interestRate) / 12)
            const principalMinusMin = principalWithInterest - minPayment;
            const principal = principalMinusMin - extraPayment;
            loanAndPrincipal.principal.push({
                date: currentDate.valueOf(),
                principal
            })
            loanAndPrincipal = iteratWithBalance(loanAndPrincipal, residualPayment);
        }

        return loanAndPrincipal
    }
```
then when used as
```ts
loanAndPrincipals = loanAndPrincipals.map(lnp => iterateWithBalace(lnp, monthlyPayment))
```
the complexity becomes just O(log2(n)) * months-projected